Add new <Code> component, to succeed <Prism>

FredKSchott opened the pull request:
## Changes

- Adds a new `<Code>` component, similar to `<Prism>`
- Powered internally by [shiki](https://shiki.matsu.io/)
- RFC: #1212 
- **Bonus:** Reimplements `Debug` component to use Code internally, removing the problem of CSS getting added to the page incorrectly! Moves Debug back to `import {Debug} from 'astro/components';` with a nice backwards-compat deprecation message for anyone still using the old import URL.

## Testing

- Tests added.

## Docs

- Docs updated.

jasikpark commented:
Could you add a link to the Shiki docs, especially the dark theme docs?

jasikpark commented on packages/astro/package.json:
```diff
@@ -17,7 +17,7 @@
     "./snowpack-plugin-jsx": "./snowpack-plugin-jsx.cjs",
     "./components": "./components/index.js",
     "./debug": "./components/Debug.astro",
-    "./components/*": "./components/*",
```
Why make this change? I'm unsure of what this changes - I guess you can no longer do `import Markdown from "astro/components/Markdown.astro";`?

FredKSchott replied:
That's correct, all are expected to now go through `import {} from 'astro/components';`

Was that ever documented? I thought that was only needed internally for the Prism component (which is why it is still added here).

jasikpark replied:
ah, I think that was only for `Debug`, no?

matthewp replied:
This was in the export map before we supported reexporting Astro components.

FredKSchott replied:
ah, maybe we just keep it then and remove in v1.0

FredKSchott replied:
added back

jasikpark reviewed (commented):
Left a comment, otherwise looks pretty great!

jasikpark commented:
I want to get the docs site setup with `<Code>` asap.. would that be via something like `remark-shiki`?

I think @natemoo-re built a theme for VSCode that we could pull in!

matthewp commented on packages/astro/components/Code.astro:
```diff
@@ -0,0 +1,21 @@
+---
+import shiki from 'shiki';
+const { code, lang = 'plaintext', theme = 'nord' } = Astro.props;
```
Can you add an interface? I think we should have those going forward. I plan on PRing a change to add interfaces for Markdown and Prism.

matthewp commented on packages/astro/components/Code.astro:
```diff
@@ -0,0 +1,21 @@
+const _html = highlighter.codeToHtml(code, lang);
+html = repairCssThemeColors(repairClassName(_html));
+---
+{html}
```
Just a question, is this adding a `<pre><code>` or are you expected to add those yourself?

jasikpark replied:
Shiki should be generating that as part of its themed html

matthewp commented:
Just a note that this PR is for the Phase 1 section fo the RFC.

jasikpark reviewed (approved):
Left a change for the `<Debug/>` overflow, but otherwise LGTM! Maybe we can get the docs light/dark mode theme setup for `<Debug/>` so Debug is astro themed? 🤔

jasikpark commented on packages/astro/components/Debug.astro:
```diff
@@ -1,306 +1,18 @@
+const key = Object.keys(Astro.props)[0];
+const value = format(Astro.props[key]);
+// Log message to console, as well as the browser.
+console.debug(`[${yellow(bold('debug'))}] ` + format({[key]: value}));
```
```suggestion
console.debug(`[${yellow(bold('debug'))}] `, {[key]: value});
```

- https://nodejs.org/api/console.html#console_console_debug_data_args
- https://nodejs.org/api/console.html#console_console_log_data_args
- https://nodejs.org/api/util.html#util_util_format_format_args

`console.debug` passes it's arguments to `util.format` so the call to format is unnecessary!

<img width="640" alt="image" src="https://user-images.githubusercontent.com/10626596/131781059-c1fd2773-102a-4309-9e90-5fe4671582ba.png">

Top: w/ my change (i didn't import `kleur/colors` in my test file

Bottom: the current code.

FredKSchott replied:
nice! great catch

jasikpark reviewed (changes requested):
Requested a change for the `console.debug` in `Debug.astro`! This is looking great!

jasikpark commented on packages/astro/components/Debug.astro:
```diff
@@ -1,306 +1,18 @@
+const key = Object.keys(Astro.props)[0];
+const value = format(Astro.props[key]);
+// Log message to console, as well as the browser.
+console.debug(`[${yellow(bold('debug'))}] `, {[key]: value});
```
Ack I didn't actually test the file, so I didn't notice `const value = format(Astro.props[key]);`.

```suggestion
console.debug(`[${yellow(bold('debug'))}] `, {[key]: Astro.props[key]});
```

We want to only have *one* round of `utils.format()` 😅

jasikpark replied:
It might also make sense to move the `util.format` call into the `<Code/>` template string, and just do `<Code lang="js" code={`// <Debug ${key}={...} />\n${format(value)}`} />`

FredKSchott replied:
yup, lgtm!

jasikpark commented:
Maybe the docs could show an example of using the debug and code components perhaps?￼

FredKSchott commented:
Talked with @matthewp and @natemoo-re today and they convinced me that the Debug change should get pulled out into a separate PR so as not to block the Phase 1 `<Code>` work that we set out to do. 

Will remove and ALSO update shiki to latest which will get our tests passing!

FredKSchott commented:
> Maybe the docs could show an example of using the debug and code components perhaps?￼

Docs already added, at `builtin-components.md`

FredKSchott commented:
Fixed the code overflow by default problem and added a new `wrap` prop to the component which lets you control overflow yourself.

```
<Code> // Default. overflow-x: auto;
<Code wrap={false}> // overflow-x: auto;
<Code wrap>  // Default: word wrap.
<Code wrap={null}> // no overflow control added. Will most likely overflow.
```

jasikpark commented:
> > Maybe the docs could show an example of using the debug and code components perhaps?￼
> 
> Docs already added, at `builtin-components.md`

I forgot to add wanting to show the outputted component w/in the page! both the source & the end result!

jasikpark commented on docs/src/pages/reference/builtin-components.md:
````diff
@@ -5,6 +5,26 @@ title: Built-In Components
+## `<Code />`
+
+```astro
+---
````
It'd be really cool to make use of https://github.com/unlight/remark-sources so that we can provide the rendered version of this code example inline with the component!
